File: src/main/java/cf/jrozen/mh/ttp/model/Individual.java
package cf.jrozen.mh.ttp.model;

import cf.jrozen.mh.ttp.GreedyKnapsackSolver;
import com.google.common.primitives.Ints;
import io.vavr.Lazy;
import io.vavr.collection.List;

import java.util.Set;

public class Individual {

    private final Context context;
    private final int[] locations;
    private final Lazy<Double> value = Lazy.of(this::valueInit);

    Individual(Context context) {
        this(context, randomGenes(context));
    }

    Double value() {
        return value.get();
    }

    private Individual(Context context, final int[] locations) {
        this.context = context;
        this.locations = locations;
    }

    private double valueInit() {
        final Set<Item> items = GreedyKnapsackSolver.chooseItems(context, this.locations);
        return context.calculate(this.locations, items);
    }

    Individual mutate() {
        final int[] cloned = locations.clone();
        for (int currInd = 0; currInd < locations.length; currInd++) {
            if (context.nextMutate()) {
                final int randomInd = context.nextIntInDims();
                int tmp = cloned[currInd];
                cloned[currInd] = cloned[randomInd];
                cloned[randomInd] = tmp;
            }
        }
        return new Individual(context, cloned);
    }

    Individual crossover(Individual that) {
        if (context.nextCrossover())
            return this.crossoverOne(that);
        else
            return this;
    }

    private Individual crossoverOne(Individual that) {
        final List<Integer> genes = List.ofAll(crossArrays(this.locations, that.locations));
        final List<Integer> fstNotUsed = List.range(0, context.problem().dimension()).removeAll(genes);
        return new Individual(context, toArray(genes.distinct().appendAll(fstNotUsed)));
    }

    private static int[] randomGenes(Context context) {
        return toArray(List.range(0, context.problem().dimension()).shuffle());
    }

    private int[] crossArrays(int[] thisGenes, int[] thatGenes) {
        final int splitAt = context.nextIntInDims();
        final int[] crossed = new int[context.problem().dimension()];
        System.arraycopy(thisGenes, 0, crossed, 0, splitAt);
        System.arraycopy(thatGenes, splitAt, crossed, splitAt, crossed.length - splitAt);
        return crossed;
    }

    private static int[] toArray(List<Integer> list) {
        return Ints.toArray(list.asJava());
    }
}

File: src/main/scala/cf/jrozen/mh/ttp/model/Individual.scala
package cf.jrozen.mh.ttp.model


import cf.jrozen.mh.ttp.utils.ValueCache
import cf.jrozen.mh.ttp.{GreedyKnapsackSolver, mutable}

import scala.util.Random

case class Individual(val locations: Array[Int])(implicit context: Context) extends Ordered[Individual] {

  //  def value: Double = valueInit
  def value: Double = valueCache.get

  private val valueCache = ValueCache[Double](() => valueInit)

  private def valueInit = {
    val items = GreedyKnapsackSolver.chooseItems(context, this.locations)
    context.calculate(this.locations, items)
  }

  @mutable
  def mutableSwap(idx1: Int, idx2: Int): Individual = {
    val x = locations(idx1)
    locations(idx1) = locations(idx2)
    locations(idx2) = x
    valueCache.invalidate
    this
  }

  def swap(idx1: Int, idx2: Int): Individual = {
    val swapped = locations.clone()
    val x = swapped(idx1)
    swapped(idx1) = swapped(idx2)
    swapped(idx2) = x
    Individual(swapped)
  }

  override def compare(that: Individual): Int = (value * 100).intValue().compareTo((that.value * 100).intValue())

  def copy() = Individual(locations)

  override def canEqual(a: Any): Boolean = a.isInstanceOf[Individual]

  override def equals(that: Any): Boolean =
    that match {
      case that: Individual => that.canEqual(this) && this.value - that.value < 0.001d
      case _ => false
    }

  override def hashCode: Int = {
    (this.value * 100).intValue()
  }

}

object Individual {
  def random(implicit context: Context): Individual =
    Individual(Random.shuffle(List.range(0, context.problem.dimension)).toArray)

  def apply(loc: Array[Int])(implicit context: Context) = new Individual(loc.clone())

}
